### PR TITLE
[5.0] Improve CMS installation code

### PR DESCRIPTION
**Pull request by Hackwar**
### Summary of Changes
The code to install the CMS is quite a mess and pretty inconsistent. The configuration sometimes is expected as an array, sometimes as an object. Sometimes the configuration is handed in from the outside, sometimes it is retrieved by the model itself from the session. The form validation code is not done like in the rest of the CMS by the controller calling the methods in the model, but the controller is calling a special method in the model, which then again does the validation calls...

Goals of this PR:
1. The configuration is explicitely given the model as method parameter. This also improves the code of the CLI installation.
2. The configuration is used as an array
3. The code is cleaned up to confirm to our regular flow of execution elsewhere in the CMS.
4. Functions are decoupled in a better way to better integrate with the CLI installer

### Link to documentations
- [x] No documentation changes for docs.joomla.org needed
- [x] No documentation changes for manual.joomla.org needed

### CODE REVIEW

**Comment by joomdonation** [comment]
I have tested this item :white_check_mark: successfully on ec4c9462fcb17729c20038d801eddce321eabbf9<br /><br />Installation works well, install additional languages work well, tried to enter wrong credentials..., all still work.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/38914">issues.joomla.org/tracker/joomla-cms/38914</a>.</sub>

File: installation/src/Model/DatabaseModel.php [hunk 1]
```diff
@@ -45,16 +45,15 @@ public function getOptions()
...
-     * @param   boolean  $select  Select the database when creating the connections.
+     * @param   array    $options  Array with db connection credentials
+     * @param   boolean  $select   Select the database when creating the connections.
      *
      * @return  DatabaseInterface|boolean  Database object on success, boolean false on failure
      *
      * @since   3.1
      */
-    public function initialise($select = true)
+    public function initialise($options, $select = true)
```
Comment (laoneo) [inline installation/src/Model/DatabaseModel.php]: Can you type hint the arguments?

File: installation/src/Model/DatabaseModel.php [hunk 2]
```diff
@@ -107,22 +106,25 @@ public function initialise($select = true)
...
      *
+     * @param   array  $options  Array of database credentials
+     *
      * @return  boolean
      *
      * @since   3.1
      * @throws  \RuntimeException
      */
-    public function createDatabase()
+    public function createDatabase($options)
```
Comment (laoneo) [inline installation/src/Model/DatabaseModel.php]: Can you type hint the argument?

File: installation/src/Model/DatabaseModel.php [hunk 3]
```diff
@@ -262,34 +264,31 @@ public function createDatabase()
...
      * Method to process the old database.
      *
+     * @param   array  $options  Array with database credentials
+     *
      * @return  boolean  True on success.
      *
      * @since   3.1
      */
-    public function handleOldDatabase()
+    public function handleOldDatabase($options)
```
Comment (laoneo) [inline installation/src/Model/DatabaseModel.php]: Can you type hint the argument?

File: installation/src/Model/DatabaseModel.php [hunk 4]
```diff
@@ -299,15 +298,16 @@ public function handleOldDatabase()
...
-     * @param   string  $schema  The SQL schema file to apply.
+     * @param   string  $schema   The SQL schema file to apply.
+     * @param   array   $options  Array with database credentials
      *
      * @return  boolean  True on success.
      *
      * @since   3.1
      */
-    public function createTables($schema)
+    public function createTables($schema, $options)
```
Comment (laoneo) [inline installation/src/Model/DatabaseModel.php]: Can you type hint the arguments?

File: installation/src/Model/SetupModel.php [hunk 5]
```diff
@@ -236,15 +206,15 @@ public function validate($data, $view = null)
...
      *
+     * @param   array  $options  Array with database credentials
+     *
      * @return  boolean
      *
      * @since   4.0.0
      * @throws  \Exception
      */
-    public function validateDbConnection()
+    public function validateDbConnection($options)
```
Comment (laoneo) [inline installation/src/Model/SetupModel.php]: Can you type hint the argument?

Comment (Hackwar) [reply installation/src/Model/SetupModel.php]: done

File: installation/src/Model/DatabaseModel.php (hunk 4 above)
Comment (Hackwar) [reply installation/src/Model/DatabaseModel.php]: done

File: installation/src/Model/DatabaseModel.php (hunk 3 above)
Comment (Hackwar) [reply installation/src/Model/DatabaseModel.php]: done

File: installation/src/Model/DatabaseModel.php (hunk 2 above)
Comment (Hackwar) [reply installation/src/Model/DatabaseModel.php]: done

File: installation/src/Model/DatabaseModel.php (hunk 1 above)
Comment (Hackwar) [reply installation/src/Model/DatabaseModel.php]: done

**Comment by Hackwar** [comment]
Codestyle has been fixed.

File: installation/src/Model/DatabaseModel.php [hunk 6]
```diff
@@ -299,15 +298,16 @@ public function handleOldDatabase()
...
-     * @param   string  $schema  The SQL schema file to apply.
+     * @param   string  $schema   The SQL schema file to apply.
+     * @param   array   $options  Array with database credentials
      *
      * @return  boolean  True on success.
      *
      * @since   3.1
      */
-    public function createTables($schema)
+    public function createTables($schema, array $options)
```
Comment (laoneo) [inline installation/src/Model/DatabaseModel.php]:
```suggestion
    public function createTables(string $schema, array $options)
```

File: installation/src/Model/DatabaseModel.php [hunk 7]
```diff
@@ -45,16 +45,15 @@ public function getOptions()
...
-     * @param   boolean  $select  Select the database when creating the connections.
+     * @param   array    $options  Array with db connection credentials
+     * @param   boolean  $select   Select the database when creating the connections.
      *
      * @return  DatabaseInterface|boolean  Database object on success, boolean false on failure
      *
      * @since   3.1
      */
-    public function initialise($select = true)
+    public function initialise(array $options, $select = true)
```
Comment (laoneo) [inline installation/src/Model/DatabaseModel.php]:
```suggestion
    public function initialise(array $options, bool $select = true)
```

File: installation/src/Controller/InstallationController.php [hunk 8]
```diff
@@ -64,19 +64,33 @@ public function dbcheck()
         $r       = new \stdClass();
         $r->view = 'setup';
 
-        // Check the form
         /** @var \Joomla\CMS\Installation\Model\SetupModel $model */
         $model = $this->getModel('Setup');
+        $data  = Factory::getApplication()->input->post->get('jform', [], 'array');
```
Comment (laoneo) [inline installation/src/Controller/InstallationController.php]: Why here not `$this->getApplication()`?

Comment (Hackwar) [reply installation/src/Controller/InstallationController.php]: Because there is not `getApplication()` in this class.

Comment (laoneo) [reply installation/src/Controller/InstallationController.php]: And what about `$this->app`?

Comment (Hackwar) [reply installation/src/Controller/InstallationController.php]: done.

**Comment by HLeithner** [comment]
pretty sure the installer get tested with beta1 much more then as this pr. thanks

**Comment by HLeithner** [comment]
It seems there are errors in the PR that have to been checked and fixed. Seems the first form is not preserved... @Hackwar I didn't discovered this while testing. @wilsonge reported it to me

**Comment by wilsonge** [comment]
So I accidentally added the wrong db admin to the form. Clicked submit and got the correct error (n.b. I didn't see the progress bar at this point). However, when fixing the error in the form and submitting again I got the error that @fancyFranci obtained and had no way to resolve it or go back. I had to refresh the page and enter all information again.

**Comment by HLeithner** [comment]
strange, my first run was also mit wrong database credentials but I got back to the prefilled form and corrected my error and installation went thru

**Comment by wilsonge** [comment]
Yeah this definitely broke stuff - but only if you clear your cookies out in the meantime because this doesn't refresh after data is updated in the session. See #41942

**Pull request merged**